File: lib/models/booking.dart
class Booking {
  final String id;
  final String futsal;
  final String futsalName;
  final DateTime date;
  final String startTime;
  final String endTime;
  final double totalPrice;
  final String status;
  final List<Map<String, dynamic>> kitRentals;

  Booking({
    required this.id,
    required this.futsal,
    required this.futsalName,
    required this.date,
    required this.startTime,
    required this.endTime,
    required this.totalPrice,
    required this.status,
    this.kitRentals = const [],
  });

  factory Booking.fromJson(Map<String, dynamic> json) {
    return Booking(
      id: json['_id'] ?? json['id'],
      futsal: json['futsal'],
      futsalName: json['futsalName'] ?? '',
      date: DateTime.parse(json['date']),
      startTime: json['startTime'],
      endTime: json['endTime'],
      totalPrice: (json['totalPrice'] ?? 0.0).toDouble(),
      status: json['status'] ?? 'pending',
      kitRentals: List<Map<String, dynamic>>.from(json['kitRentals'] ?? []),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'futsal': futsal,
      'futsalName': futsalName,
      'date': date.toIso8601String(),
      'startTime': startTime,
      'endTime': endTime,
      'totalPrice': totalPrice,
      'status': status,
      'kitRentals': kitRentals,
    };
  }

  Booking copyWith({
    String? id,
    String? futsal,
    String? futsalName,
    DateTime? date,
    String? startTime,
    String? endTime,
    double? totalPrice,
    String? status,
    List<Map<String, dynamic>>? kitRentals,
  }) {
    return Booking(
      id: id ?? this.id,
      futsal: futsal ?? this.futsal,
      futsalName: futsalName ?? this.futsalName,
      date: date ?? this.date,
      startTime: startTime ?? this.startTime,
      endTime: endTime ?? this.endTime,
      totalPrice: totalPrice ?? this.totalPrice,
      status: status ?? this.status,
      kitRentals: kitRentals ?? this.kitRentals,
    );
  }
} 

File: backend/models/Booking.js
const mongoose = require('mongoose');

const bookingSchema = new mongoose.Schema({
    user: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    futsal: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'Futsal',
        required: true
    },
    date: {
        type: Date,
        required: true
    },
    startTime: {
        type: String,
        required: true
    },
    endTime: {
        type: String,
        required: true
    },
    totalPrice: {
        type: Number,
        required: true
    },
    status: {
        type: String,
        enum: ['pending', 'confirmed', 'cancelled', 'completed'],
        default: 'pending'
    },
    kitRentals: [{
        kit: {
            type: mongoose.Schema.Types.ObjectId,
            ref: 'Kit',
            required: true
        },
        quantity: {
            type: Number,
            required: true,
            min: 1,
            validate: {
                validator: Number.isInteger,
                message: '{VALUE} is not an integer value'
            }
        },
        price: {
            type: Number,
            required: true,
            min: 0
        }
    }],
    createdAt: {
        type: Date,
        default: Date.now
    }
});

// Index for efficient querying of bookings by date and futsal
bookingSchema.index({ futsal: 1, date: 1 });

// Static method to get available time slots for a court on a specific date
bookingSchema.statics.getAvailableSlots = async function(futsalId, date) {
    const bookings = await this.find({
        futsal: futsalId,
        date: date,
        status: { $in: ['pending', 'confirmed'] }
    });
    
    return bookings;
};

// Static method to check if a time slot is available
bookingSchema.statics.isSlotAvailable = async function(futsalId, date, startTime, endTime) {
    const conflictingBooking = await this.findOne({
        futsal: futsalId,
        date: date,
        status: { $in: ['pending', 'confirmed'] },
        $or: [
            { startTime: { $lt: endTime }, endTime: { $gt: startTime } }
        ]
    });
    
    return !conflictingBooking;
};

module.exports = mongoose.model('Booking', bookingSchema); 